### ISSUE TITLE
Bad count of HTML pages on main replay page

### ISSUE THREAD

**Issue by machawk1** [oduwsdl/ipwb#556]
Some HTML pages may not have the precise content-type response header of "text/html", e.g., http://acid.matkelly.com/ has `text/html; charset=UTF-8`. On the replay page, we are relying on this value to be precisely the string `text/html`, which is incorrect and unreliable.

HTML pages are hidden in the interface by default because of this and the count is incorrect. The "hiding" is the result of

```
ul#uriList li[data-mime="text/html"] {background-color: white; display: block;}
```
in webui.css, which might need to be `*=` instead of simply `=`.

Further, ```if j['mime_type'] == 'text/html':``` in replay.py may need to be something like ```if 'text/html' in j['mime_type']:```

**Comment by machawk1** [comment]
Above suggestions applied in issue-556 branch. More testing needed.

**Comment by machawk1** [comment]
A caveat to which I would like your input, @ibnesayeed.

In a WARC I have for acid.matkelly.com there are these (select) WARC response records:

* 1 resource with `Content-Type: text/html; charset=UTF-8` at time _A_
* Another HTML resource with `Content-Type: text/html` at time _B_
* A request for a .js HTTP URI at time _C_ with `Content-Type: text/html; charset=iso-8859-1` whose body consists solely of a `HTTP/1.1 301 Moved Permanently`
* A request for the previous resource at time _C_ but on HTTPS but `application/javascript`
* A request for a .png HTTP URI at time _D_ with `Content-Type: text/html; charset=iso-8859-1` whose body consists solely of a `HTTP/1.1 301 Moved Permanently`
* A request for the previous resource at time _D_ but on HTTPS but `Content-Type: image/png`

Using our current logic of just checking for `content-type`, we report 4 "HTML pages", two of which are 301 redirects. Should we also ensure that the status code is not a redirect as well?

Do you anticipate any side-effects of the count of HTML pages if we add this additional condition on top of the match of the content-type starting with `text/html`?

**Comment by ibnesayeed** [comment]
Ideally we should only list URI-Rs, linked to their latest URI-M, with a count of mementos for each URI-R which will be linked to the URI-M listing page. And yes, filtering only pages with successful response code is a good idea. The intent is to identify the best candidates to surface.

**Comment by machawk1** [comment]
Why is the latest "the best" in this case?

**Comment by ibnesayeed** [comment]
> Why is the latest "the best" in this case?

That's a convention from the TimeGate resolution where it would default to the latest memento when no `Accept-Datetime` is supplied. However, if something else makes more sense then we can go for that. "The best" candidate identification was in appreciation of the one with `OK` response, not necessarily for the latest one.

**Comment by machawk1** [comment]
@ibnesayeed Ok. Those improvements seems like they should be in a separate issue. I know we had discussed this list of URI-Rs+datetime in the past but could not find a ticket for the improvement.

I will go forward with excluding the mementos with a 3XX response code, despite their Content-Type, for now on the main page list. Since they are still historical captures, they can still be present in the comprehensive list (which we will eventually refine anyway to be an abbreviated set of _the best_ mementos), but ought not be shown in the default list.

**Comment by machawk1** [comment]
There still exists an issue relevant with this ticket wherein the memento of the redirect of _URI-M_ at _t_ is overwriting the same _URI-M_ at _t_ with a non-redirect status code. This is within the JSON in the JS of the landing page `<head>`.

**Comment by machawk1** [comment]
8bdec52 excludes the 3XXs but the count is still incorrect (should be 4, not 2) based on examples in https://github.com/oduwsdl/ipwb/issues/556#issuecomment-422220979 and issue described in https://github.com/oduwsdl/ipwb/issues/556#issuecomment-422388332.

**Comment by machawk1** [comment]
When the `let uris = {}` is populated in the replay page JS, the `mime` attribute for a URI-R is implemented to be a string whereas the `datetimes` attribute is an array. This is typically the case but a `mime` is not guaranteed to be stable over time.

As an unrelated note, we are not accounting for mementos of the XHTML variety that may have MIMEs like `application/xhtml+xml`. 🔨  🕳🕳🐹🕳

**Comment by machawk1** [comment]
Again with the 🔨 🕳🕳🐹🕳:

The scheme is stripped off in the surt process, so URI-Ms with http-->https at the same datetime are coalesced into the same JSON object in the replay UI JS.

**Comment by ibnesayeed** [comment]
> The scheme is stripped off in the surt process, so URI-Ms with http→https at the same datetime are coalesced into the same JSON object in the replay UI JS.

This issue is important different enough to be discussed in a separate ticket. There are a few ways to handle this, but they will require some indexing and replay logic changes.

**Comment by ibnesayeed** [comment]
> When the `let uris = {}` is populated in the replay page JS, the `mime` attribute for a URI-R is implemented to be a string whereas the `datetimes` attribute is an array. This is typically the case but a `mime` is not guaranteed to be stable over time.

Essentially, we want a data structure that looks something like:

```json
{
	"example.com": [{
		"datetime": 20161222134829,
		"status": 200,
		"mime": "text/html"
	}, {
		"datetime": 20171125141852,
		"status": 302,
		"mime": "text/plain"
	}],
	"example.org": [{
		"datetime": 20150312133330,
		"status": 200,
		"mime": "text/html"
	}]
}
```
